app/src/main/java/com/keylesspalace/tusky/util/FocalPointUtil.kt

```diff
@@ -0,0 +1,161 @@
+/* Copyright 2018 Jochem Raat <jchmrt@riseup.net>
+ *
+ * This file is a part of Tusky.
+ *
+ * This program is free software; you can redistribute it and/or modify it under the terms of the
+ * GNU General Public License as published by the Free Software Foundation; either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * Tusky is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
+ * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+ * Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with Tusky; if not,
+ * see <http://www.gnu.org/licenses>. */
+
+package com.keylesspalace.tusky.util
+
+import android.graphics.Matrix
+import android.widget.ImageView
+
+import com.keylesspalace.tusky.entity.Attachment.Focus
+import com.squareup.picasso.Callback
+
+/**
+ * Calculates the image matrix needed to maintain the correct cropping for image views based on
+ * their focal point.
+ *
+ * The purpose of this class is to make sure that the focal point information on media
+ * attachments are honoured. This class uses the custom matrix option of android ImageView's to
+ * customize how the image is cropped into the view.
+ *
+ * See the explanation of focal points here:
+ *    https://github.com/jonom/jquery-focuspoint#1-calculate-your-images-focus-point
+ */
+object FocalPointUtil {
+    /**
+     * Create the appropriate matrix for the given parameters.
+     *
+     * How it works is using the following steps:
+     *   - First we determine if the image is too wide or too tall for the view size. If it is
+     *   too wide, we need to crop it horizontally and scale the height to fit the view
+     *   exactly. If it is too tall we need to crop vertically and scale the width to fit the
+     *   view exactly.
+     *   - Then we determine what translation is needed to get the focal point in view. We
+     *   prefer to get the focal point at the center of the preview. However if that would
+     *   result in some part of the preview being empty, we instead align the image so that it
+     *   fills the view, but still the focal point is always in view.
+     *
+     * @param viewWidth The width of the imageView.
+     * @param viewHeight The height of the imageView
+     * @param imageWidth The width of the actual image
+     * @param imageHeight The height of the actual image
+     * @param focus The focal point to focus
+     * @param mat The matrix to use, this matrix is reset() and then updated with the new
+     * configuration. We reuse the old matrix to prevent unnecessary allocations.
+     *
+     * @return The matrix which correctly crops the image
+     */
+    fun updateFocalPointMatrix(viewWidth: Float,
```

Comment:
I think returning it is a little bit redundant but maybe it doesn't hurt to be have more flexible interface

Reply:
On the other hand, it's less clear that matrix is mutated if it's also returned

Reply:
I think you're right, it's clearer that it mutates the matrix if it doesn't return it. I've changed it to not return the matrix.